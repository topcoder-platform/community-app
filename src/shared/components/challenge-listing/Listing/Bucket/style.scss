@import "~styles/mixins";

.bucket {
  background: $tc-gray-neutral-light;
  margin: 20px 0;
}

.view-more,
.view-more:active,
.view-more:focus,
.view-more:hover,
.view-more:visited {
  background: $tc-gray-neutral-light;
  border-top: 1px solid $tc-gray-10;
  border-bottom: 1px solid $tc-gray-10;
  border-left: 0;
  border-radius: 0;
  border-right: 0;
  color: $tc-gray-70;
  cursor: pointer;
  display: block;
  font: 300 13px Arial, sans-serif;
  outline: none;
  padding: 12px 0;
  text-align: center;
  text-transform: none;
  width: 100%;
}

.view-more:focus {
  border: 3px solid #79a3ef;
  border-radius: 4px;
}
