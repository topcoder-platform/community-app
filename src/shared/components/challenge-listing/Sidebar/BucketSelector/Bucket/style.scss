@import "~styles/mixins";

.active {
  background: $tc-gray-10;
  cursor: default;
  font-weight: 600;

  @include xs-to-sm {
    background: $tc-white;
  }
}

.bucket {
  color: $tc-black;
  border-radius: 2 * $corner-radius;
  cursor: pointer;
  padding: 0 2 * $base-unit 0 3 * $base-unit;

  @include xs-to-sm {
    font-size: 15px;
    padding: 2px 0;
  }
}

.right {
  color: $tc-gray-70;
  float: right;
  position: static;
  font-weight: 700;
}

.errorMsg {
  color: $tc-red;
  line-height: 13px;
  padding-bottom: 6px;
  display: none;
}

.active .errorMsg {
  display: block;
}
