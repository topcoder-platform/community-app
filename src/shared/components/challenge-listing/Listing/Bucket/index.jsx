/**
 * A single bucket of challenges.
 */

/* global document */

import _ from 'lodash';
import PT from 'prop-types';
import qs from 'qs';
import React from 'react';
import Sort from 'utils/challenge-listing/sort';
import SortingSelectBar from 'components/SortingSelectBar';
import Waypoint from 'react-waypoint';
import { getFilterFunction } from 'utils/challenge-listing/filter';
import CardPlaceholder from '../../placeholders/ChallengeCard';
import ChallengeCard from '../../ChallengeCard';
import './style.scss';

const COLLAPSED_SIZE = 10;

export default function Bucket({
  bucket,
  bucketId,
  challenges,
  challengesUrl,
  expanded,
  expand,
  filterState,
  keepPlaceholders,
  loading,
  loadMore,
  newChallengeDetails,
  openChallengesInNewTabs,
  prizeMode,
  selectChallengeDetailsTab,
  selectedCommunityId,
  setFilterState,
  setSort,
  sort,
  userHandle,
  expandedTags,
  expandTag,
}) {
  const filter = getFilterFunction(bucket.filter);
  const activeSort = sort || bucket.sorts[0];

  const sortedChallenges = _.clone(challenges);
  sortedChallenges.sort(Sort[activeSort].func);

  const bucketQuery = qs.stringify({
    bucket: bucketId,
    communityId: selectedCommunityId || undefined,
    filter: filterState,
  }, { encodeValuesOnly: true });

  let expandable = false;
  const filteredChallenges = [];
  for (let i = 0; i < sortedChallenges.length; i += 1) {
    if (filter(sortedChallenges[i])) {
      filteredChallenges.push(sortedChallenges[i]);
    }
    if (!expanded && filteredChallenges.length >= COLLAPSED_SIZE) {
      expandable = true;
      break;
    }
  }

  if (!filteredChallenges.length && !loadMore) return null;

  const cards = filteredChallenges.map(item => (
    <ChallengeCard
      challenge={item}
      challengesUrl={challengesUrl}
      newChallengeDetails={newChallengeDetails}
      onTechTagClicked={tag => setFilterState({ tags: [tag] })}
      openChallengesInNewTabs={openChallengesInNewTabs}
      prizeMode={prizeMode}
      key={item.id}
      selectChallengeDetailsTab={selectChallengeDetailsTab}
      userHandle={userHandle}
      expandedTags={expandedTags}
      expandTag={expandTag}
    />
  ));

  const placeholders = [];
  if (loading || keepPlaceholders) {
    for (let i = 0; i < 8; i += 1) {
      placeholders.push(<CardPlaceholder id={i} key={i} />);
    }
  }

  return (
    <div styleName="bucket">
      <SortingSelectBar
        onSelect={setSort}
        options={
          bucket.sorts.map(item => ({
            label: Sort[item].name,
            value: item,
          }))
        }
        title={bucket.name}
        value={{
          label: Sort[activeSort].name,
          value: activeSort,
        }}
      />
      {cards}
      {
        !expandable && loadMore && !loading ? (
          <Waypoint onEnter={loadMore} />
        ) : null
      }
      {placeholders}
      {
        (expandable || loadMore) && !keepPlaceholders && !loading && !expanded ? (
          <a
            href={`${challengesUrl}?${bucketQuery}`}
            onClick={(event) => {
              expand();
              document.body.scrollTop = 0;
              document.documentElement.scrollTop = 0;
              event.preventDefault();
            }}
            role="button"
            styleName="view-more"
            tabIndex={0}
          >View more challenges</a>
        ) : null
      }
    </div>
  );
}

Bucket.defaultProps = {
  expanded: false,
  expand: _.noop,
  keepPlaceholders: false,
  loading: false,
  loadMore: null,
  newChallengeDetails: false,
  openChallengesInNewTabs: false,
  sort: null,
  userHandle: '',
  expandedTags: [],
  expandTag: null,
};

Bucket.propTypes = {
  bucket: PT.shape().isRequired,
  bucketId: PT.string.isRequired,
  expanded: PT.bool,
  expand: PT.func,
  challenges: PT.arrayOf(PT.shape()).isRequired,
  challengesUrl: PT.string.isRequired,
  filterState: PT.shape().isRequired,
  keepPlaceholders: PT.bool,
  loading: PT.bool,
  loadMore: PT.func,
  newChallengeDetails: PT.bool,
  openChallengesInNewTabs: PT.bool,
  prizeMode: PT.string.isRequired,
  selectChallengeDetailsTab: PT.func.isRequired,
  selectedCommunityId: PT.string.isRequired,
  setFilterState: PT.func.isRequired,
  setSort: PT.func.isRequired,
  sort: PT.string,
  userHandle: PT.string,
  expandedTags: PT.arrayOf(PT.number),
  expandTag: PT.func,
};
