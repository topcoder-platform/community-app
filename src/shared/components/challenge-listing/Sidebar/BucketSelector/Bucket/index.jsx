/**
 * Regular sidebar row.
 */

import _ from 'lodash';
import { challenge as challengeUtils } from 'topcoder-react-lib';
import PT from 'prop-types';
import { connect } from 'react-redux';
import React from 'react';
import './style.scss';

const Filter = challengeUtils.filter;

function Bucket({
  active,
  bucket,
  challenges,
  disabled,
  onClick,
  allActiveChallengesLoaded,
  meta,
}) {
  let countEl;
  if (!bucket.hideCount && !disabled) {
    const filter = Filter.getFilterFunction(bucket.filter);
    let count;
    if (allActiveChallengesLoaded) {
      count = challenges.filter(filter).length;
    } else {
      switch (bucket.name) {
        case 'All Challenges':
          count = meta.allChallengesCount;
          break;
        case 'My Challenges':
          count = meta.myChallengesCount;
          break;
        case 'Open for registration':
          count = meta.openChallengesCount;
          break;
        case 'Ongoing challenges':
          count = meta.ongoingChallengesCount;
          break;
        default:
      }
    }
    countEl = (
      <span styleName="right">
        {count}
      </span>
    );
  }

  const error = Boolean(bucket.error) && (
    <div styleName="errorMsg">
      {bucket.error}
    </div>
  );

  if (active) {
    return (
      <div styleName="active bucket">
        {bucket.name}
        {countEl}
        {error}
      </div>
    );
  }

  return (
    <div
      onClick={disabled ? _.noop : onClick}
      onKeyPress={e => (e.key === 'Enter' ? onClick() : null)}
      role="button"
      styleName="bucket"
      tabIndex={0}
    >
      {bucket.name}
      {countEl}
      {error}
    </div>
  );
}

Bucket.defaultProps = {
  active: false,
  disabled: false,
  onClick: _.noop,
  meta: {},
};

Bucket.propTypes = {
  active: PT.bool,
  bucket: PT.shape({
    hideCount: PT.bool,
    name: PT.string.isRequired,
    error: PT.string,
    filter: PT.any,
  }).isRequired,
  challenges: PT.arrayOf(PT.shape).isRequired,
  disabled: PT.bool,
  onClick: PT.func,
  meta: PT.shape(),
  allActiveChallengesLoaded: PT.bool.isRequired,
};

const mapStateToProps = (state) => {
  const cl = state.challengeListing;
  return {
    allActiveChallengesLoaded: cl.allActiveChallengesLoaded,
    meta: cl.meta,
  };
};

const BucketContainer = connect(
  mapStateToProps,
)(Bucket);

export default BucketContainer;
