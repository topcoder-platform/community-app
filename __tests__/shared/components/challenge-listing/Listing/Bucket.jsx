import React from 'react';
// import ReactDOM from 'react-dom';
import _ from 'lodash';
import Renderer from 'react-test-renderer/shallow';
// import TU from 'react-dom/test-utils';
import Bucket from 'components/challenge-listing/Listing/Bucket';
import reduxStoreFactory from 'redux-mock-store';
import { Provider } from 'react-redux';
import { StaticRouter } from 'react-router-dom';
// import { Tag } from 'topcoder-react-ui-kit';

const store = reduxStoreFactory()();

const expand = jest.fn();
const loadMore = jest.fn();
const setFilterState = jest.fn();
const setSort = jest.fn();

const mockDatas = [{
  bucket: 'all',
  expanded: true,
  expand,
  challenges: [
    {
      id: '1',
      status: 'b',
      prizes: [1200, 600],
      totalPrize: 1800,
      users: {},
      type: 'Challenge',
      tags: [],
    },
    {
      id: '2',
      status: 'a',
      technologies: ['a', 'b', 'c'],
      tags: ['a', 'b', 'c'],
      allPhases: [{
        phaseType: 'Registration',
        phaseStatus: 'Open',
      }],
      currentPhases: ['Registration'],
      track: 'Development',
      type: 'Challenge',
      events: [{ eventName: 'Submit' }],
      prizes: [1200, 600],
      totalPrize: 1800,
      users: {},
    }],
  challengeTypes: [
    {
      name: 'Challenge',
      abbreviation: 'CH',
    }, {
      name: 'First2Finish',
      abbreviation: 'F2F',
    }, {
      name: 'Task',
      abbreviation: 'TSK',
    }, {
      name: 'Marathon Match',
      abbreviation: 'MM',
    },
  ],
  loading: false,
  loadMore,
  setFilterState,
  setSort,
  sort: '',
}];

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();

  _.forEach(mockDatas, (data) => {
    renderer.render((
      <Provider store={store}>
        <StaticRouter context={{}}>
          <Bucket {...data} />
        </StaticRouter>
      </Provider>
    ));
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});

// class Wrapper extends React.Component {
//   componentDidMount() {}

//   render() {
//     return (
//       <Provider store={store}>
//         <StaticRouter context={{}}>
//           <Bucket {...this.props} />
//         </StaticRouter>
//       </Provider>
//     );
//   }
// }

// const instance = TU.renderIntoDocument((<Wrapper {...mockDatas[0]} />));

// test('setFilterState', () => {
//   const matches = TU.scryRenderedComponentsWithType(instance, Tag);
//   expect(matches).toHaveLength(3);
// TU.Simulate.click(ReactDOM.findDOMNode(matches[0]));
// expect(setFilterState).toHaveBeenCalledTimes(1);
// TU.Simulate.click(ReactDOM.findDOMNode(matches[1]));
// expect(setFilterState).toHaveBeenCalledTimes(2);
// TU.Simulate.click(ReactDOM.findDOMNode(matches[2]));
// expect(setFilterState).toHaveBeenCalledTimes(3);
// });
