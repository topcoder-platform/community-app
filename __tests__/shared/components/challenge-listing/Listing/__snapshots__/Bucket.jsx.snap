// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/"
>
  <Bucket
    bucket={
      Object {
        "filter": Object {
          "status": "ac",
        },
        "sorts": Array [
          "current-phase",
        ],
      }
    }
    challenges={
      Array [
        Object {
          "id": "1",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "b",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "allPhases": Array [
            Object {
              "phaseStatus": "Open",
              "phaseType": "Registration",
            },
          ],
          "currentPhases": Array [
            "Registration",
          ],
          "events": Array [
            Object {
              "eventName": "Submit",
            },
          ],
          "id": "2",
          "isDataScience": false,
          "prize": Array [
            1200,
            600,
          ],
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "subTrack": "CODE",
          "technologies": "a, b, c",
          "totalPrize": 1800,
          "track": "DEVELOP",
          "users": Object {},
        },
      ]
    }
    expand={[Function]}
    expandTag={null}
    expanded={true}
    expandedTags={Array []}
    keepPlaceholders={false}
    loadMore={[Function]}
    loading={false}
    newChallengeDetails={false}
    openChallengesInNewTabs={false}
    setFilterState={[Function]}
    setSort={[Function]}
    sort=""
    userHandle=""
  />
</StaticRouter>
`;

exports[`Matches shallow shapshot 2`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/"
>
  <Bucket
    bucket={
      Object {
        "filter": Object {
          "status": "ac",
        },
        "sorts": Array [
          "current-phase",
        ],
      }
    }
    challenges={
      Array [
        Object {
          "id": "1",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "b",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "2",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "3",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "4",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "5",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "6",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "7",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "8",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "9",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "10",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
        Object {
          "id": "11",
          "prizes": Array [
            1200,
            600,
          ],
          "status": "a",
          "totalPrize": 1800,
          "users": Object {},
        },
      ]
    }
    expand={[Function]}
    expandTag={null}
    expanded={false}
    expandedTags={Array []}
    keepPlaceholders={false}
    loadMore={[Function]}
    loading={true}
    newChallengeDetails={false}
    openChallengesInNewTabs={false}
    setFilterState={[Function]}
    setSort={[Function]}
    sort=""
    userHandle=""
  />
</StaticRouter>
`;

exports[`Matches shallow shapshot 3`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/"
>
  <Bucket
    bucket={
      Object {
        "filter": Object {
          "status": "e",
        },
        "sorts": Array [
          "current-phase",
        ],
      }
    }
    challenges={Array []}
    expand={[Function]}
    expandTag={null}
    expanded={false}
    expandedTags={Array []}
    keepPlaceholders={false}
    loadMore={null}
    loading={true}
    newChallengeDetails={false}
    openChallengesInNewTabs={false}
    setFilterState={[Function]}
    setSort={[Function]}
    sort=""
    userHandle=""
  />
</StaticRouter>
`;
